### PR TITLE
Sparse interpolation optimizations

### PR DESCRIPTION
**Pull request by mo-g**
Two changes here.

The first adds a single housekeep to the compute_interp_weights function, which reduces some in-process memory usage.

The second replaces a cartesian product using itertools, with one based solely on numpy. On my sample data, this took 3.27ms instead of 846ms, and used ~13.5MiB ram instead of ~126.5MiB.

Limitations of the new function are that all the points must be of the same datatype, which I _think_ is irrelevant in our usage? (Correct me if I'm wrong...)

### CODE REVIEW

**Comment by pelson** [comment]
:+1: if you fix the pep8 problems.

File: lib/iris/analysis/_scipy_interpolate.py [hunk 1]
```diff
@@ -226,6 +226,7 @@ def compute_interp_weights(self, xi, method=None):
                 n_indices = np.ravel_multi_index(corner_indices, shape,
                                                  mode='wrap')
                 col_indices[i::n_src_values_per_result_value] = n_indices
+                del(n_indices)
```
Comment (ajdawson) [inline lib/iris/analysis/_scipy_interpolate.py]: Would this be simpler if you just assigned the `np.ravel` call to the slice of `col_indices` directly without the intermediate `n_indices`?

**Comment by mo-g** [comment]
Tests passed, rebasing.

File: lib/iris/analysis/_interpolation.py [hunk 2]
```diff
@@ -472,6 +471,18 @@ def _interpolated_dtype(self, dtype):
...
+    def _product(dimensions):
+        # Calculate cartesian product using numpy instead of itertools.
+        # inspired by http://stackoverflow.com/posts/11146645/revisions
+        ndim = len(dimensions)
+        shape = [len(points) for points in dimensions] + [ndim]
+        product = np.empty(shape, dtype=dimensions[0].dtype)
+        mesh = np.ix_(*dimensions)
+        for index, points in enumerate(mesh):
+            product[..., index] = points
+        return product.reshape(-1, ndim)
```
Comment (bjlittle) [inline lib/iris/analysis/_interpolation.py]: @mo-g Some test coverage for this would be nice :wink:

**Comment by mo-g** [comment]
Don't merge this, please. I've got some more results on time use and memory use that I'd like to analyse.

**Comment by pelson** [comment]
> Don't merge this, please. I've got some more results on time use and memory use that I'd like to analyse.

Shame. I could happily have merged this, and then merged another PR....

**Comment by rhattersley** [comment]
Nudge!

> I've got some more results on time use and memory use that I'd like to analyse.

Is this change sometimes detrimental? Or is just that it doesn't always help?

**Comment by rhattersley** [comment]
>> I've got some more results on time use and memory use that I'd like to analyse.
> 
> Is this change sometimes detrimental? Or is just that it doesn't always help?

@mo-g - it'd be really useful to understand how this approach doesn't help (or if it does!)

**Comment by mo-g** [comment]
From internal note.. 09/Jul/15 15:49

> Used Tehuti to look at three branches of Iris from a perspective of memory and processing time:

> https://github.com/mo-g/iris/tree/master-pre-sparse-interpolation
> https://github.com/mo-g/iris/tree/sparse-unoptimized
> https://github.com/mo-g/iris/tree/sparse-interpolation-optimizations

> The branch 'sparse-interpolation-optimizations' is a set of two changes I added based on profiling-
> by-hand with psutil to analyse different lines of code for memory usage and speed. The first change
> removes the assignment of a variable that is only used once. The second replaces an iterative
> cartesian product with one based on meshing. It was intended that these would mitigate some of the
> memory costs of the sparse interpolation; however they are not a consistent improvement - as was
> shown when using Tehuti to profile them. As such, the pull request has been held for now.

**Comment by mo-g** [comment]
Ran some Tehuti tests that @dkillick wrote a while back.

```
* regrid (tiny, to rotated)
* regrid from limited area (large - tiny)
* regrid (tiny, from rotated)
* rotated regrid (tiny)
* rotated regrid (tiny - large)
* regrid from limited area to rotated (large, to rotated)
* simple regrid (tiny)
* rotated regrid (large)
* vertical levels regrid (large, from rotated)
* regrid from rotated limited area (large, from rotated)
* regrid to limited area (tiny - large)
* regrid from limited area (tiny - large)
* vertical levels rotated regrid (large)
* regrid from limited area (large, nan extrap)
* simple regrid (large)
* regrid (large, to rotated)
* vertical levels regrid (tiny - tiny)
* vertical levels regrid to limited area (large)
* regrid from limited area (large - large)
* regrid to limited area (large - large)
* regrid (large, from rotated)
* regrid to limited area (tiny, from rotated)
* regrid from limited area (tiny - tiny)
* rotated regrid to limited area (tiny - tiny)
* vertical levels regrid from limited area (large)
* regrid to limited area (tiny - tiny)
* regrid to limited area (tiny, to rotated)
* vertical levels regrid (large - large)
* vertical levels regrid (large, to rotated)
* regrid to limited area (large - tiny)
* regrid from limited area (large, mask extrap)
* rotated regrid from limited area (large)
```
weighted_optimization

Attached image shows the percentage change (multiplied by the actual size of the optimized output, to give some weighting) in peak memory usage against time for each of the above; plus counts of how many are in each of the four 'bins'.

I think this encourages merging?
@rhattersley @pelson?

File: lib/iris/analysis/_interpolation.py [hunk 3]
```diff
@@ -472,6 +471,18 @@ def _interpolated_dtype(self, dtype):
             result = np.result_type(_DEFAULT_DTYPE, dtype)
         return result
 
+    @staticmethod
+    def _product(dimensions):
```
Comment (pelson) [inline lib/iris/analysis/_interpolation.py]: Is this "orthogonal_point_coordinates"?

File: lib/iris/analysis/_interpolation.py [hunk 4]
```diff
@@ -472,6 +471,18 @@ def _interpolated_dtype(self, dtype):
...
+        # Calculate cartesian product using numpy instead of itertools.
+        # inspired by http://stackoverflow.com/posts/11146645/revisions
+        ndim = len(dimensions)
+        shape = [len(points) for points in dimensions] + [ndim]
+        product = np.empty(shape, dtype=dimensions[0].dtype)
```
Comment (pelson) [inline lib/iris/analysis/_interpolation.py]: I'm not sure we can assume that the dtype is consistent. See https://github.com/SciTools/iris/pull/1717/files#diff-07b1c9cb6b22e14b7f314da7af652aafR554 which may produce different dtypes for each dimension.

Comment (pelson) [inline lib/iris/analysis/_interpolation.py]: At the least, perhaps we can check, and only optimise if it is...

**Comment by rhattersley** [comment]
@mo-g - thanks for the graph. :+1: 

So... mostly small <5% time changes, but a few more significant changes. It'd be great it you could list these 3 or 4 tests.

Similarly, for the memory use, is there a pattern to which tests result in a non-trivial increase?

**Comment by mo-g** [comment]
Here are the absolute values (I had the script printing these before I got funky with MPL.)

Tim Increase: regrid from limited area (large - tiny) 0.67% 0.00s
Tim Increase: regrid from limited area to rotated (large, to rotated) 0.28% 0.00s
Mem Increase: simple regrid (tiny) 0.07% 0.33MB
Tim Increase: simple regrid (tiny) 2.87% 0.00s
Mem Increase: vertical levels regrid (large, from rotated) 0.74% 13.93MB
Tim Increase: regrid to limited area (tiny - large) 1.28% 0.00s
Tim Increase: regrid from limited area (tiny - large) 0.20% 0.00s
Mem Increase: vertical levels rotated regrid (large) 0.74% 13.93MB
Mem Increase: vertical levels regrid to limited area (large) 0.74% 13.93MB
Tim Increase: vertical levels regrid to limited area (large) 9.91% 0.12s
Tim Increase: regrid from limited area (large - large) 0.25% 0.00s
Tim Increase: regrid to limited area (tiny, from rotated) 0.24% 0.00s
Tim Increase: rotated regrid to limited area (tiny - tiny) 4.60% 0.01s
Mem Increase: vertical levels regrid from limited area (large) 0.74% 13.93MB
Mem Increase: vertical levels regrid (large - large) 0.74% 13.93MB
Mem Increase: vertical levels regrid (large, to rotated) 0.74% 13.93MB

**Comment by mo-g** [comment]
Here are all the >5% changes:
```
Tim Increase: regrid (tiny, to rotated) -36.% -0.00s
Tim Increase: rotated regrid (tiny) -7.8% -0.00s
Tim Increase: rotated regrid (tiny - large) -21.% -0.32s
Tim Increase: rotated regrid (large) -5.5% -0.03s
Tim Increase: vertical levels regrid (large, from rotated) -7.0% -0.15s
Tim Increase: vertical levels regrid to limited area (large) 9.91% 0.12s
Tim Increase: vertical levels regrid from limited area (large) -5.7% -0.06s
```

**Comment by rhattersley** [comment]
> Tim Increase: vertical levels regrid to limited area (large) 9.91% 0.12s

Shame, I was hoping this would be from a small regrid, not a non-trivial "global @ (2000, 1800) -> regional (900, 600)". :unamused:

**Comment by rhattersley** [comment]
Looks like this would benefit from a better understanding of what the memory & runtime trends are with respect to source & destination resolutions. @mo-g - since you're not in a position to do that investigation right now I don't mind if you want to close this PR. @pelson - any preference on what to do next?

**Comment by pelson** [comment]
Think we need to close this one as has been on pause quite some time.

**Pull request closed without merging**